sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/EmptyCollectionsShouldNotBeEnumerated.cs

```diff
@@ -0,0 +1,342 @@
+            {
+                // Only one-dimensional arrays can be empty, others are indeterminate, because this code becomes ugly
+                if (arrayCreation.Type.RankSpecifiers.Count != 1 ||
+                    arrayCreation.Type.RankSpecifiers[0].Sizes.Count != 1)
+                {
+                    return null;
+                }
+
+                var size = arrayCreation.Type.RankSpecifiers[0].Sizes[0] as LiteralExpressionSyntax;
+
+                return size?.Token.ValueText == "0"
+                    ? CollectionCapacityConstraint.Empty
+                    : null;
+            }
+
+            private CollectionCapacityConstraint GetCollectionConstraint(IMethodSymbol constructor)
+            {
+                // Default constructor, or constructor that specifies capacity means empty collection,
+                // otherwise do not apply constraint because we cannot be sure what has been passed
+                // as arguments.
+                var defaultCtorOrCapacityCtor = !constructor.Parameters.Any()
+                    || constructor.Parameters.Count(p => p.IsType(KnownType.System_Int32)) == 1;
+
+                return defaultCtorOrCapacityCtor ? CollectionCapacityConstraint.Empty : null;
+            }
+
+            private CollectionCapacityConstraint GetInitializerConstraint(InitializerExpressionSyntax initializer)
+            {
+                if (initializer == null)
+                {
+                    return null;
+                }
+
+                return initializer.Expressions.Count == 0
+                    ? CollectionCapacityConstraint.Empty // No items added through the initializer
+                    : CollectionCapacityConstraint.NotEmpty;
+            }
+
+            private static ProgramState RemoveCollectionConstraintsFromArguments(InvocationExpressionSyntax invocation,
+                ProgramState programState)
+            {
+                // Remove the collection constraints from all arguments of the invocation expression
+
+                var values = invocation.ArgumentList.Arguments.Select((node, index) =>
+                {
+                    SymbolicValue value;
+                    programState.ExpressionStack.Pop(out value);
+                    return value;
+                }).ToList();
+
+                foreach (var value in values)
+                {
+                    programState = value.RemoveConstraint(CollectionCapacityConstraint.Empty, programState);
+                }
+
+                return programState;
+            }
+
+            private static bool IsDictionarySetItem(ElementAccessExpressionSyntax elementAccess) =>
+                (elementAccess.GetSelfOrTopParenthesizedExpression().Parent as AssignmentExpressionSyntax)
+                    ?.Left.RemoveParentheses() == elementAccess;
+
+            private static bool IsCollectionConstructor(IMethodSymbol constructorSymbol) =>

```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Consider using more general type 'Microsoft.CodeAnalysis.ISymbol' instead of 'Microsoft.CodeAnalysis.IMethodSymbol'. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS3242)